### PR TITLE
Fix Camera2D is not working inside a MainScreenEditorPlugin

### LINKED ISSUE #33931: Camera2D in a viewport while running as an editor plugin doesn't work
**Godot version:**
Godot v3.1.2

**OS/device including version:**
Windows 10

**Issue description:**
Whenever running as an editor plugin, Camera2D nodes won't work in a separate viewport.

**Steps to reproduce:**
1. Make an EditorPlugin.
2. Make a scene with a Viewport, Sprite and a Camera2D node.
3. Add a script to make the Camera2D node move around.
3. Run the editor plugin and try to move the camera around.
4. The Camera2D node doesn't work at all.

**Minimal reproduction project:**
https://github.com/Flairieve/GraphicsEditor/

### PR DESCRIPTION
**Pull request by WhalesState**
- Production edit: This closes https://github.com/godotengine/godot/issues/33931.

### CODE REVIEW

**Comment by WhalesState** [comment]
just one more fix, draw screen/limits/drag margins are not updating if is editing in editor and inside a custom viewport, how to fix it ? by checking `if (is_editing_in_editor() || custom_viewport && (custom_viewport->get_viewport() == get_tree()->get_edited_scene_root().get_viewport()))` ?

**Comment by WhalesState** [comment]
I think removing _is_editing_in_editor() checks inside the 3 methods will fix it since it only calls queue_redraw() which will draw them in the next process frame, testing now and i hope it works as expected.

**Comment by WhalesState** [comment]
After testing I have found that this has many benefits:

1- Camera2D works as expected in Editor edited scene root if not inside a SubViewport else it will directly update the SubViewport preview from the camera if it's enabled , and if disabled it will has no effect but only draw screen/limits/drag_margins will not be shown
2- inside a MainScreenEditorPlugin it will work as if inside game, which is amazing like Pixelorama for example can work directly inside a MainScreenEditorPlugin and get released as a godot addon!

my only issue now is that adding 2 lines after include stops the checks with error and removing the extra lines generates error in one linux check, so for now i will only include one line and will wait if anyone can help me making draw screen/limits/drag margins be visible if inside a SubViewport and not enabled

File: scene/2d/camera_2d.cpp [hunk 1]
```diff
@@ -33,6 +33,11 @@
 #include "core/config/project_settings.h"
 #include "scene/main/window.h"
 
+
```
Comment (AThousandShips) [inline scene/2d/camera_2d.cpp]: Need to remove this line for style, multiple empty lines are not permitted

Recommend using code formatting, see [here](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#doc-code-style-guidelines)

File: scene/2d/camera_2d.cpp [hunk 2]
```diff
@@ -434,7 +439,7 @@ void Camera2D::_make_current(Object *p_which) {
 }
 
 void Camera2D::_update_process_internal_for_smoothing() {
-	bool is_not_in_scene_or_editor = !(is_inside_tree() && Engine::get_singleton()->is_editor_hint());
+	bool is_not_in_scene_or_editor = !(is_inside_tree() && (Engine::get_singleton()->is_editor_hint() && _is_editing_in_editor()));
```
Comment (AThousandShips) [inline scene/2d/camera_2d.cpp:441]:
```suggestion
	bool is_not_in_scene_or_editor = !(is_inside_tree() && Engine::get_singleton()->is_editor_hint() && _is_editing_in_editor());
```
Not needed

**Comment by YuriSizov** [comment]
The approach overall makes sense, though I haven't reviewed or tested the changes yet.

One thing that you'll have to do eventually for this to be merged is to squash your commits into one. Please, make sure that the final commit has a short but descriptive message (the title of this PR is a good option). See [this documentation](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase), if you need help with squashing.

File: scene/2d/camera_2d.cpp [hunk 3]
```diff
@@ -272,7 +276,7 @@ void Camera2D::_notification(int p_what) {
 
 #ifdef TOOLS_ENABLED
 		case NOTIFICATION_DRAW: {
-			if (!is_inside_tree() || !Engine::get_singleton()->is_editor_hint()) {
+			if (!is_inside_tree() || !(Engine::get_singleton()->is_editor_hint() && _is_editing_in_editor())) {
```
Comment (AThousandShips) [inline scene/2d/camera_2d.cpp:279]:
```suggestion
			if (!is_inside_tree() || !Engine::get_singleton()->is_editor_hint() && !_is_editing_in_editor()) {
```

File: scene/2d/camera_2d.cpp [hunk 4]
```diff
@@ -843,4 +848,4 @@ Camera2D::Camera2D() {
 
 	set_notify_transform(true);
 	set_hide_clip_children(true);
-}
+}
```
Comment (AThousandShips) [inline scene/2d/camera_2d.cpp:851]: This should not have been removed, restore the newline here

**Comment by WhalesState** [comment]
i want to cry, i have done this ` git reset --soft HEAD~13 && git commit -m "Fix Camera2D is not working inside a MainScreenEditorPlugin"` then forced a push

**Comment by AThousandShips** [comment]
You can still reset to the last commit `5c4f7b2` to restore your changes

**Comment by WhalesState** [comment]
fixed from inside vscode with undo and force push

**Comment by AThousandShips** [comment]
One or more of your checks are incorrect, causing free access to some null pointer, you need to test and figure this out locally

**Comment by WhalesState** [comment]
maybe this variable `is_editing` ? in line :575 
`Size2 Camera2D::_get_camera_screen_size() const {
	bool is_editing = get_tree()->get_edited_scene_root() && get_viewport() == get_tree()->get_edited_scene_root()-
}`

**Comment by AThousandShips** [comment]
Yes you need to check that `get_tree` doesn't return `null`

File: scene/2d/camera_2d.cpp [hunk 5]
```diff
@@ -568,7 +572,8 @@ Point2 Camera2D::get_camera_screen_center() const {
 
 Size2 Camera2D::_get_camera_screen_size() const {
 	// special case if the camera2D is in the root viewport
-	if (Engine::get_singleton()->is_editor_hint() && get_viewport()->get_parent_viewport() == get_tree()->get_root()) {
+	bool is_editing = get_tree() && get_tree()->get_edited_scene_root() && get_viewport() == get_tree()->get_edited_scene_root()->get_viewport();
```
Comment (AThousandShips) [inline scene/2d/camera_2d.cpp:575]: Isn't this equivalent to the method above? Should use that method I'd say and update it

**Comment by WhalesState** [comment]
`get_viewport()->get_parent_viewport() == get_tree()->get_root()` 
I think it's not, this will return the root if the Camera2D is inside a SubViewport() so it's like
`get_viewport()->get_parent()->get_viewport()` so if the Camera2D is inside Node2D under root this will be like getting the viewport of the root

**Comment by WhalesState** [comment]
but it's usefull to check if the `Camera2D` is inside a `Subviewport` and not `enabled` then we can draw screen/limits/drag_margins to the `SubViewport` if `_is_editing_in_editor()` returns `true`

**Comment by WhalesState** [comment]
```
Viewport *Viewport::get_parent_viewport() const {
	ERR_READ_THREAD_GUARD_V(nullptr);
	ERR_FAIL_COND_V(!is_inside_tree(), nullptr);
	if (!get_parent()) {
		return nullptr; //root viewport
	}

	return get_parent()->get_viewport();
}
```

**Comment by AThousandShips** [comment]
Please add to your previous comments if you're just adding new information, making a bunch of comments one after another will ping people and clutter up

**Comment by WhalesState** [comment]
should i also add the changes to godot-cpp repo ?

**Comment by AThousandShips** [comment]
Not here, this is due to some memory issues, not related to this PR

**Comment by WhalesState** [comment]
I have spent some more time cleaning the code and removing complex expressions, based on warnings from CodeScene vscode plugin and also testing it, it's easier for reviewing now.

**Comment by WhalesState** [comment]
changed _is_editing_in_editor() to be a const member function

**Comment by AThousandShips** [comment]
I'd say no, it'd clutter up the class and they are only used in one place, it also makes you have to jump around to find where those functions are

**Comment by WhalesState** [comment]
ok, just let me know if you can find any piece of code that needs editing or any useless checks since I'm new to c++, Thank you ^^

**Comment by AThousandShips** [comment]
You've included an unrelated commit from another PR of yours in this

**Comment by WhalesState** [comment]
i didn't touch this branch :) was working on dev1, i will protect it from force push

**Comment by KoBeWi** [comment]
Do you have some minimal project where this can be tested? The linked issue doesn't have one.

**Comment by WhalesState** [comment]
I have this addon, create a new project, create `addons` folders then extract it inside and enable it in project settings
[insider_camera.zip](https://github.com/godotengine/godot/files/12286589/insider_camera.zip)

File: scene/2d/camera_2d.cpp [hunk 6]
```diff
@@ -33,20 +33,21 @@
 #include "core/config/project_settings.h"
 #include "scene/main/window.h"
 
+bool Camera2D::_is_editing_in_editor() const {
```
Comment (KoBeWi) [inline scene/2d/camera_2d.cpp:36]: There is `is_part_of_edited_scene()` Node method, which does the same.

File: scene/2d/camera_2d.cpp [hunk 7]
```diff
@@ -453,15 +462,25 @@ void Camera2D::make_current() {
...
 	}
 }
 
 bool Camera2D::is_current() const {
-	if (viewport && !(custom_viewport && !ObjectDB::get_instance(custom_viewport_id))) {
+	if (!viewport) {
+		return false;
+	}
+
+	if (!custom_viewport && !ObjectDB::get_instance(custom_viewport_id)) {
```
Comment (KoBeWi) [inline scene/2d/camera_2d.cpp:480]: Is this some refactoring? It changes logic and looks wrong. If anything, it should be
```suggestion
	if (!custom_viewport || ObjectDB::get_instance(custom_viewport_id)) {
```

File: scene/2d/camera_2d.cpp [hunk 8]
```diff
@@ -453,15 +462,25 @@ void Camera2D::make_current() {
 
 void Camera2D::clear_current() {
 	ERR_FAIL_COND(!is_current());
-	if (viewport && !(custom_viewport && !ObjectDB::get_instance(custom_viewport_id)) && viewport->is_inside_tree()) {
+
+	if (!viewport || !viewport->is_inside_tree()) {
+		return;
+	}
+
+	if (!custom_viewport && !ObjectDB::get_instance(custom_viewport_id)) {
```
Comment (KoBeWi) [inline scene/2d/camera_2d.cpp:470]: Same here.

**Comment by WhalesState** [comment]
Thank you ^^ I've updated the code and tested inside Editor Plugin, this also fixes the other issue where white color was drawing black! I will push the changes now ^^

image

**Comment by WhalesState** [comment]
Also I'm sorry about the refactor, just have got many warnings from CodeScene plugin about complex expressions.

**Comment by WhalesState** [comment]
should i change `is_part_of_edited_scene()` to `Node2D::is_part_of_edited_scene()` ?

**Comment by WhalesState** [comment]
can i make a local function like this

```c++

bool _is_editing_in_editor() const {
#ifdef TOOLS_ENABLED
    return is_part_of_edited_scene();
#endif
    return false;
}

```

**Comment by AThousandShips** [comment]
Sure, though it should be;
```cpp
bool _is_editing_in_editor() const {
#ifdef TOOLS_ENABLED
    return is_part_of_edited_scene();
#else
    return false;
#endif // TOOLS_ENABLED
}
```
Or you'll get errors by returning multiple times

**Comment by WhalesState** [comment]
one more const function isn't a big deal for this fix, right ?

```c++

bool Camera2D::_is_editing_in_editor() const {
#ifndef TOOLS_ENABLED
	return false;
#else
	return is_part_of_edited_scene();
#endif // !TOOLS_ENABLED
}

```

File: scene/2d/camera_2d.cpp [hunk 9]
```diff
@@ -33,20 +33,21 @@
 #include "core/config/project_settings.h"
 #include "scene/main/window.h"
 
+bool Camera2D::_is_editing_in_editor() const {
+#ifndef TOOLS_ENABLED
+	return false;
+#else
+	return is_part_of_edited_scene();
+#endif // !TOOLS_ENABLED
```
[the remark this answers is not in the document]
Comment (AThousandShips) [reply scene/2d/camera_2d.cpp:41]: I'd swap the order back, easier to read

**Comment by YuriSizov** [comment]
If you have "if - else", then there is no reason to use `#ifndef`, just use `#ifdef`. Having a wrapper method is fine here, yes.

**Comment by KoBeWi** [comment]
There is a bug, the camera borders no longer follow the node:

https://github.com/godotengine/godot/assets/2223172/e38e6da0-279b-48e1-8324-904ff63f498e

**Comment by WhalesState** [comment]
fixed

https://github.com/godotengine/godot/assets/53877170/65ac95c7-e1eb-4345-874a-54b5db81fbb5

File: scene/2d/camera_2d.cpp [hunk 10]
```diff
@@ -250,8 +251,10 @@ void Camera2D::_notification(int p_what) {
 			add_to_group(group_name);
 			add_to_group(canvas_group_name);
 
-			if (!Engine::get_singleton()->is_editor_hint() && enabled && !viewport->get_camera_2d()) {
-				make_current();
+			if (!_is_editing_in_editor()) {
+				if (enabled && !viewport->get_camera_2d()) {
```
Comment (KoBeWi) [inline scene/2d/camera_2d.cpp:255]: Why did you split the condition? If you think the line is too long you can break it, but the inner condition is redundant.

**Review by KoBeWi — APPROVED** [review]

File: scene/2d/camera_2d.cpp [hunk 11]
```diff
@@ -453,15 +458,25 @@ void Camera2D::make_current() {
...
 bool Camera2D::is_current() const {
-	if (viewport && !(custom_viewport && !ObjectDB::get_instance(custom_viewport_id))) {
+	if (!viewport) {
+		return false;
+	}
+
+	if (!custom_viewport || ObjectDB::get_instance(custom_viewport_id)) {
 		return viewport->get_camera_2d() == this;
 	}
+
```
Comment (KoBeWi) [inline scene/2d/camera_2d.cpp:479]: I think this blank line is redundant here.

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Comment by WhalesState** [comment]
> Thanks! And congrats for your first merged Godot contribution 🎉

Thank you, I'm so happy ^^. have a nice day 🎉.

**Pull request merged**